core/src/main/java/org/ldaptive/pool/BlockingConnectionPool.java

```diff
@@ -103,24 +103,33 @@ public Connection getConnection()
       // block here until create occurs without locking the whole pool
       // if the pool is already maxed or creates are failing,
       // block until a connection is available
-      checkOutLock.lock();
       try {
-        boolean b = true;
-        poolLock.lock();
+        if (!checkOutLock.tryLock(blockWaitTime.toMillis(), TimeUnit.MILLISECONDS)) {
+          logger.debug("Block time of {} exceeded, throwing exception", blockWaitTime);
+          throw new BlockingTimeoutException(
+            "Block time of " + blockWaitTime + " exceeded waiting for check out on pool " + getName() +
+              " with max size of " + getMaxPoolSize());
+        }
         try {
-          logger.trace("create connection in pool of size {}", available.size() + active.size());
-          if (available.size() + active.size() == getMaxPoolSize()) {
-            logger.trace("pool at maximum size, create not allowed");
-            b = false;
+          boolean b = true;
+          poolLock.lock();
+          try {
+            logger.trace("create connection in pool of size {}", available.size() + active.size());
+            if (available.size() + active.size() == getMaxPoolSize()) {
+              logger.trace("pool at maximum size, create not allowed");
+              b = false;
+            }
+          } finally {
+            poolLock.unlock();
```

Comment:
I believe you have created a potential race condition here by unlocking the pool prior to calling `createActiveConnection`:

1. Thread 1 attempts to create a new connection
2. Thread 2 attempts to create a new connection
4. Thread 1 locks the pool and is below max
5. Thread 2 blocks on obtaining the pool lock
6. Thread 1 determines size is below threshold, sets `b=true`, releases lock and invokes `createActiveConnection`
7. Thread 2 unblocks and obtains the lock and checks pool size
8. Thread 1 blocks on obtaining the pool lock in `createActiveConnection`
9. Thread 2 determines pool size is below limit and sets `b=true` and releases lock
10. Thread 1 obtains lock and creates new connection
11. Thread 2 invokes `createActiveConnection`

Upon completion of Thread 2's invocation, the pool MAY be over limit.

Reply:
Thread 2 should be waiting on the `checkOutLock` in this scenario before even getting to #2. This specific block of code is necessary so the entire pool isn't blocked waiting for new connections to be created.

Reply:
Believe you're correct. I guess this proves the point in my summary: analyzing changes in locking behavior are tricky.